aten/src/ATen/cpu/vec256/functional.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+#include "vec256.h"
+
+namespace at { namespace vec256 {
+namespace {
+
+template <int64_t size>
+inline int64_t _leftover(int64_t x, int64_t y) {
+  if (x + size > y)
+    return y - x;
+  return size;
+}
+
+} // namespace
+
+template <typename scalar_t>
+inline scalar_t reduce_all(
+    Vec256<scalar_t> (*vec_fun)(Vec256<scalar_t>&, Vec256<scalar_t>&),
+    scalar_t* data,
+    int64_t size,
+    scalar_t ident) {
+  Vec256<scalar_t> acc_vec(ident);
+  for (int64_t d = 0; d < size; d += Vec256<scalar_t>::size) {
+    Vec256<scalar_t> value(ident);
+    int64_t leftover = _leftover<Vec256<scalar_t>::size>(d, size);
+    value.load(data + d, leftover);
+    acc_vec = vec_fun(acc_vec, value);
+  }
+  scalar_t acc_arr[Vec256<scalar_t>::size];
+  acc_vec.store(acc_arr);
+
+  for (int64_t i = Vec256<scalar_t>::size / 2; i >= 1; i = i / 2) {
+    scalar_t acc_arr_first[Vec256<scalar_t>::size];
+    scalar_t acc_arr_second[Vec256<scalar_t>::size];
+    for (int64_t j = 0; j < i; j++) {
+      acc_arr_first[j] = acc_arr[j];
+      acc_arr_second[j] = acc_arr[j + i];
+    }
```

Comment:
I can't see why do you need to scatter the element into two arrays. Seems to me like you could just do (are misaligned accesses the problem?)
```cpp
vec_first.load(acc_arr, i);
vec_second.load(acc_arr + i, i);
```

Anyway, I thought of a different way of summing the values, that would only use vector instructions. You can find an example for SSE [here](https://stackoverflow.com/a/10930706/3343043).